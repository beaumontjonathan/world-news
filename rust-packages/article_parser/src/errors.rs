use chrono::ParseError as ChronoParseError;
use htmlescape::DecodeErr;
use scraper::Selector;
use serde_json::Error as SerdeJsonError;

#[derive(Debug)]
pub enum InternalParseError {
  UnknownError(&'static str),
  HtmlStructureError(HtmlStructureError),
  JsonParseError(JsonParseError),
  InvalidDateTime(ChronoParseError),
  HtmlDecodeErr(DecodeErr),
}

#[derive(Debug)]
pub enum JsonParseErrorCause {
  InvalidLdJson,
  InvalidDataAttrJson,
}

#[derive(Debug)]
pub struct JsonParseError {
  pub cause: JsonParseErrorCause,
  pub serde_error: SerdeJsonError,
}

#[derive(Debug)]
pub enum HtmlErrorCause {
  MissingElement,
  NonUniqueElement,
  MissingAttribute(String),
}

#[derive(Debug)]
pub struct HtmlStructureError {
  pub cause: HtmlErrorCause,
  pub selector: &'static Selector,
}

impl InternalParseError {
  pub fn html(error: HtmlStructureError) -> InternalParseError {
    InternalParseError::HtmlStructureError(error)
  }

  pub fn json(error: JsonParseError) -> InternalParseError {
    InternalParseError::JsonParseError(error)
  }

  pub fn ld_json(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidLdJson,
      serde_error,
    })
  }

  pub fn json_data_attr(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidDataAttrJson,
      serde_error,
    })
  }

  pub fn unknown(s: &'static str) -> InternalParseError {
    InternalParseError::UnknownError(s)
  }

  pub fn chrono(err: ChronoParseError) -> InternalParseError {
    InternalParseError::InvalidDateTime(err)
  }

  pub fn decode_html(err: DecodeErr) -> InternalParseError {
    InternalParseError::HtmlDecodeErr(err)
  }
}

impl From<ChronoParseError> for InternalParseError {
  fn from(err: ChronoParseError) -> Self {
    InternalParseError::chrono(err)
  }
}

impl From<DecodeErr> for InternalParseError {
  fn from(err: DecodeErr) -> Self {
    InternalParseError::decode_html(err)
  }
}
