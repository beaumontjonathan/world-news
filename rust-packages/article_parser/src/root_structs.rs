use chrono::{DateTime, Utc};
use serde::Serialize;
use std::fmt::Debug;

#[derive(Debug, Serialize)]
pub struct Image {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct Video {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
pub enum Publisher {
  BBC,
  DailyMail,
  Independent,
  Metro,
  Mirror,
  Sky,
  Sun,
  Guardian,
}

use Publisher::*;

impl From<String> for Publisher {
  fn from(publisher: String) -> Self {
    match publisher.to_lowercase().as_str() {
      "bbc" => BBC,
      "dailymail" => DailyMail,
      "independent" => Independent,
      "metro" => Metro,
      "mirror" => Mirror,
      "sky" => Sky,
      "sun" => Sun,
      "theguardian" | "guardian" => Guardian,
      _ => panic!("Invalid publisher type"),
    }
  }
}

#[derive(Debug, Serialize)]
pub struct NewsArticleContent {
  pub headline: String,
  pub twitter_headline: Option<String>,
  pub description: String,
  pub thumbnail: Image,
  pub categories: Vec<String>,
  pub images: Vec<Image>,
  pub videos: Vec<Video>,
  pub body: String,
  pub date_updated: DateTime<Utc>,
  pub date_published: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct VideoArticleContent {
  pub something: String,
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, Serialize)]
pub enum ParsedPageContent {
  NewsArticle(NewsArticleContent),
  VideoArticle(VideoArticleContent),
}

#[derive(Debug, Serialize)]
pub struct PageMeta {
  pub publisher: Publisher,
  pub url: String,
  pub date_parsed: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct ParsedPage {
  pub meta: PageMeta,
  pub content: ParsedPageContent,
}
