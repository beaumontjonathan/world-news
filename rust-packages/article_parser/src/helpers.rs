use crate::errors::{HtmlErrorCause, HtmlStructureError, InternalParseError};
// use chrono::{DateTime, Utc};
use crate::root_structs::ParsedPageContent;
use lazy_static::lazy_static;
use regex::Regex;
use scraper::{ElementRef, Html, Selector};
// use std::fmt::{self, Debug, Display, Formatter};
pub trait ToOwnedOrNone {
  fn to_owned_or_none(self) -> Option<String>;
}

impl<'a> ToOwnedOrNone for Option<&'a str> {
  fn to_owned_or_none(self) -> Option<String> {
    self.and_then(|mut s| {
      s = s.trim();
      if s.is_empty() {
        None
      } else {
        Some(String::from(s))
      }
    })
  }
}
// pub trait NoneIfEmpty
// where
//   Self: Sized,
// {
//   fn none_if<F>(self, callback: F) -> Option<Self>
//   where
//     F: Fn(&Self) -> bool,
//   {
//     if callback(&self) {
//       None
//     } else {
//       Some(self)
//     }
//   }
// }
// impl<T> NoneIfEmpty for T {}

pub type ParseResult = Result<ParsedPageContent, InternalParseError>;

pub trait FixHtmlWhitespace {
  fn fix_html_whitespace(self) -> Self;
}

fn fix_whitespace(text: String) -> String {
  lazy_static! {
    static ref RE: Regex = Regex::new(r#"\s+"#).unwrap();
  }
  RE.replace_all(&text, " ").into_owned()
}

impl FixHtmlWhitespace for String {
  fn fix_html_whitespace(self) -> String {
    fix_whitespace(self)
  }
}

pub trait AttrValueOrNone {
  fn attr_value_or_none(&self, attr: &str) -> Option<String>;
}

impl<'a> AttrValueOrNone for ElementRef<'a> {
  fn attr_value_or_none(&self, attr: &str) -> Option<String> {
    self.value().attr(attr).to_owned_or_none()
  }
}

pub trait FullFixedHtmlText {
  fn full_fixed_html_text(self) -> String;
}

impl<'a> FullFixedHtmlText for ElementRef<'a> {
  fn full_fixed_html_text(self) -> String {
    self
      .text()
      .map(|text| text.trim().to_string())
      .collect::<Vec<String>>()
      .join(" ")
      .fix_html_whitespace()
      .trim()
      .to_owned()
  }
}

pub trait JoinWithNewline {
  fn join_with_newline(self) -> String;
}

impl<T> JoinWithNewline for T
where
  T: Iterator<Item = String>,
{
  fn join_with_newline(self) -> String {
    self.collect::<Vec<String>>().join("\n").trim().to_owned()
  }
}

pub trait UniqueElementRef<'a> {
  fn unique(self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause>;
}

impl<'a, T> UniqueElementRef<'a> for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn unique(mut self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause> {
    match (self.next(), self.next()) {
      (None, _) => Err(HtmlErrorCause::MissingElement),
      (_, Some(_)) => Err(HtmlErrorCause::NonUniqueElement),
      (Some(el_ref), None) => Ok(el_ref),
    }
  }
}

pub trait SelectHelpers {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError>;
}

impl SelectHelpers for Html {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

impl SelectHelpers for scraper::element_ref::ElementRef<'_> {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

pub trait FormattedText {
  fn formatted_text(self) -> String;
}

impl<'a, T> FormattedText for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn formatted_text(self) -> String {
    self.map(|p| p.full_fixed_html_text()).join_with_newline()
  }
}

pub trait ChildrenHelpers {
  fn children_first(&self) -> Option<ElementRef>;
  fn children_all(&self) -> Vec<ElementRef>;
}

impl ChildrenHelpers for scraper::element_ref::ElementRef<'_> {
  fn children_first(&self) -> Option<ElementRef> {
    self.children_all().into_iter().next()
  }

  fn children_all(&self) -> Vec<ElementRef> {
    self.children().filter_map(ElementRef::wrap).collect()
  }
}
