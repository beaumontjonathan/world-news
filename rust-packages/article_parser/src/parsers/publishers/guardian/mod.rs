use crate::errors::{HtmlErrorCause, HtmlStructureError, InternalParseError};
use crate::helpers::{
  AttrValueOrNone, ChildrenHelpers, FullFixedHtmlText, JoinWithNewline, ParseResult, SelectHelpers,
};
use crate::root_structs::{Image, NewsArticleContent, ParsedPageContent, Video};

use crate::my_selectors;
use chrono::{DateTime, Utc};
use scraper::{Html, Selector};

my_selectors! {
  ARTICLE_VIDEO_PRECHECK => "body > div > article[id=\"article\"][itemtype=\"http://schema.org/VideoObject\"].content--media--video",
  TWITTER_TITLE_SELECTOR => "head > meta[name=\"twitter:text:title\"]",
  DATE_UPDATED_SELECTOR => "meta[property=\"article:modified_time\"]",
  DATE_PUBLISHED_SELECTOR => "meta[property=\"article:published_time\"]",
  HEADER => "div.content__main-column > header",
  HEADLINE => "h1.content__headline",
  DESCRIPTION => "div.content__standfirst",
  FIGURE_HEADER => "figure.media-primary",
  IMG => "img",
  FIGCAPTION => "figcaption.caption",
  ARTICLE_CONTENT => "div.content__article-body",
  ELEMENT_IMAGE => "figure.element-image",
  ELEMENT_INTERACTIVE => "figure.element-interactive",
  VIDEO_FIGURE => "figure.element[data-atom-type=\"media\"]",
  YOUTUBE_IFRAME => "div.youtube-media-atom__iframe",
}

type Result<T> = std::result::Result<T, InternalParseError>;

pub fn parse(document: Html) -> ParseResult {
  // PRECHECKS
  match document.select(&ARTICLE_VIDEO_PRECHECK).next() {
    None => Ok(()),
    Some(_) => Err(InternalParseError::unknown(
      "Article type: video article, not yet supported",
    )),
    // Some(_) => Err(InternalParseError::precondition(
    //   "Article type: video article, not yet supported",
    // )),
  }?;

  let header = document.select_unique(&HEADER)?;
  let article_content = document.select_unique(&ARTICLE_CONTENT)?;

  // HEADLINE
  let headline = {
    let h1s = header
      .select(&HEADLINE)
      .map(|h1| h1.full_fixed_html_text())
      .collect::<Vec<_>>();
    match h1s.len() {
      0 => Err(InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector: &HEADLINE,
      })),
      1 => Ok(h1s.into_iter().next().unwrap()),
      _ => {
        if h1s.windows(2).all(|w| w[0] == w[1]) {
          Ok(h1s.into_iter().next().unwrap())
        } else {
          Err(InternalParseError::unknown(
            "Unknown headline type - too many <h1>s with differing text",
          ))
        }
      }
    }
  }?;

  // TWITTER HEADLINE
  let twitter_headline = document
    .select_unique_attr(&TWITTER_TITLE_SELECTOR, "content")
    .ok();

  // DESCRIPTION
  let description = header.select_unique(&DESCRIPTION)?.full_fixed_html_text();

  // THUMBNAIL
  // todo - implement
  let thumbnail = {
    let figure = header.select_unique(&FIGURE_HEADER)?;
    let img = figure.select_unique(&IMG)?;
    let alt = img.attr_value_or_none("alt");
    let url = img.attr_value_or_none("src");
    let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
    Image { alt, url, caption }
  };

  // CATEGORIES
  // TODO - implement
  let categories: Vec<String> = vec![];

  // IMAGES
  let images = article_content
    .children_all()
    .into_iter()
    .filter(|el| ELEMENT_IMAGE.matches(&el) || ELEMENT_INTERACTIVE.matches(&el))
    .map(|figure| {
      if ELEMENT_IMAGE.matches(&figure) {
        let img = figure.select_unique(&IMG)?;
        let alt = img.attr_value_or_none("alt");
        let url = img.attr_value_or_none("src");
        let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
        Ok(Image { alt, url, caption })
      } else if ELEMENT_INTERACTIVE.matches(&figure) {
        // unimplemented!();
        Ok(Image {
          alt: None,
          url: None,
          caption: None,
        })
      } else {
        panic!("how are you here?");
      }
    })
    .collect::<Result<Vec<_>>>()?;

  // VIDEOS
  let videos = article_content
    .children_all()
    .into_iter()
    .filter(|figure| VIDEO_FIGURE.matches(&figure))
    .map(|figure| {
      let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
      let youtube_id = figure.select_unique_attr(&YOUTUBE_IFRAME, "data-asset-id")?;
      let url = Some(format!("https://www.youtube.com/watch?v={}", youtube_id));

      Ok(Video {
        alt: None,
        url,
        caption,
      })
    })
    .collect::<Result<Vec<_>>>()?;

  // BODY
  let body = article_content
    .children_all()
    .into_iter()
    .filter_map(|el| {
      if el.value().name() == "p" || el.value().name() == "ul" {
        Some(el.full_fixed_html_text())
      } else {
        None
      }
    })
    .collect::<Vec<String>>()
    .into_iter()
    .join_with_newline();

  // DATE_UPDATED
  let date_updated = document.select_unique_attr(&DATE_UPDATED_SELECTOR, "content")?;
  let date_updated = DateTime::parse_from_rfc3339(&date_updated)?.with_timezone(&Utc);

  //DATE_PUBLISHED
  let date_published = document.select_unique_attr(&DATE_PUBLISHED_SELECTOR, "content")?;
  let date_published = DateTime::parse_from_rfc3339(&date_published)?.with_timezone(&Utc);

  Ok(ParsedPageContent::NewsArticle(NewsArticleContent {
    headline,
    twitter_headline,
    description,
    thumbnail,
    categories,
    images,
    videos,
    body,
    date_updated,
    date_published,
  }))
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::test_helpers::helpers::read_html_doc;
  use insta::assert_debug_snapshot;

  #[test]
  fn test_file_01() {
    let document = read_html_doc("test-01.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_02() {
    let document = read_html_doc("test-02.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_03() {
    let document = read_html_doc("test-03.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  // todo - error nicely?
  #[test]
  fn test_file_04() {
    let document = read_html_doc("test-04.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_05() {
    let document = read_html_doc("test-05.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_06() {
    let document = read_html_doc("test-06.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_07() {
    let document = read_html_doc("test-07.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_08() {
    let document = read_html_doc("test-08.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }
}
