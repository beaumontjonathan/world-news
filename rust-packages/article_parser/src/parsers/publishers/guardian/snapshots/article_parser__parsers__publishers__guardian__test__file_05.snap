---
source: src/parsers/publishers/guardian/mod.rs
expression: article
---
Ok(
    NewsArticle(
        NewsArticleContent {
            headline: "The isolated tribes at risk of illness from Amazon missionaries",
            twitter_headline: None,
            description: "As evangelical Christians use their influence with Brazil’s government to cast their net ever wider, indigenous people vulnerable to common diseases face a growing threat",
            thumbnail: Image {
                alt: Some(
                    "A young Marubo boy, Metsisi Marubo, carries water from the Rio Itui toward the village of Rio Novo",
                ),
                url: Some(
                    "https://i.guim.co.uk/img/media/782fdf1a3863246878f621c6d19273cf5d6731ee/54_820_5111_3067/master/5111.jpg?width=300&quality=85&auto=format&fit=max&s=13d195b820dcbe6a081f4cd260969249",
                ),
                caption: Some(
                    "A young Marubo boy, Metsisi Marubo, carries water from the Rio Itui toward the village of Rio Novo. Medical experts say indigeneos people are at high risk from coronavirus. Photograph: Gary Calton/The Observer",
                ),
            },
            categories: [],
            images: [
                Image {
                    alt: Some(
                        "Ricardo Lopes Dias",
                    ),
                    url: Some(
                        "https://i.guim.co.uk/img/media/ccd4d31204bf1b98127ff726b9dfb883a333490f/0_0_774_540/master/774.jpg?width=300&quality=85&auto=format&fit=max&s=4e7266005ac80829aee5465c70775dd7",
                    ),
                    caption: Some(
                        "Former evangelical missionary Ricardo Lopes Dias is in charge of Brazil’s isolated indigenous tribes. Photograph: Mário Vilela/Funai",
                    ),
                },
                Image {
                    alt: Some(
                        "A man holding a Bible prays during a Christian event called March for Jesus in Brasilia, Brazil",
                    ),
                    url: Some(
                        "https://i.guim.co.uk/img/media/a06a7781dad6922241d4f70917d7f402a059c54c/0_0_5760_3840/master/5760.jpg?width=300&quality=85&auto=format&fit=max&s=bc4358b848c73e48195592a583bf88c2",
                    ),
                    caption: Some(
                        "A man holding a Bible prays during a Christian event called March for Jesus in Brasilia that was attended by Brazil’s president. Photograph: Eraldo Peres/AP",
                    ),
                },
                Image {
                    alt: Some(
                        "Aerial view of a village in the Yanomami indigenous territory in the north of Brazil, close to the Venezuelan border",
                    ),
                    url: Some(
                        "https://i.guim.co.uk/img/media/a7ced993ddfa53d39e4d570ca02fcf1fba46b261/21_0_2695_1617/master/2695.jpg?width=300&quality=85&auto=format&fit=max&s=c207ff5fbafda067940dadab85164882",
                    ),
                    caption: Some(
                        "A village in the Yanomami indigenous territory in the north of Brazil, close to the Venezuelan border. At least three groups of Yanomami are uncontacted. Photograph: Guilherme Gnipper Trevisan/Hutukara",
                    ),
                },
                Image {
                    alt: Some(
                        "Young Korubo children (from left) Tamo Lala, Tupa, Tëpi and Visa are seen on the Amazonian Indigenous reserve of Vale do Javeri",
                    ),
                    url: Some(
                        "https://i.guim.co.uk/img/media/f247f615f80db6943bff1b7ab3febd9c6feebb58/0_0_6720_4480/master/6720.jpg?width=300&quality=85&auto=format&fit=max&s=dcecf96dbd29a6d534279c16a87dc2c7",
                    ),
                    caption: Some(
                        "Young Korubo children (from left) Tamo Lala, Tupa, Tëpi and Visa are seen on the Amazonian Indigenous reserve of Vale do Javeri. Photograph: Gary Calton/The Observer",
                    ),
                },
            ],
            videos: [],
            body: "A radical group of evangelical Christian missionaries set on converting every last tribe on Earth has raised fears that deadly diseases – and even the coronavirus – will spread in the Brazilian Amazon. The group has based its newly bought helicopter right beside a reserve with the world’s highest concentration of isolated indigenous groups, who have little resistance to common illnesses.\nThere are more than 100 isolated indigenous groups in Brazil, all highly vulnerable to common diseases such as measles and flu, and 16 of them live in the same reserve in the Javari Valley, a vast, remote area the size of Austria. Covid-19 could wipe out any of them.\nVideos posted in 2018 by the US-based missionary organisation Ethnos360 – formerly called New Tribes Mission – aimed to raise money to buy the helicopter, saying it would be used to reach “new people groups”. The Brazilian government official now in charge of isolated and recently contacted tribes at the government’s indigenous agency, Funai, is a former New Tribes missionary .\nThe official, anthropologist Ricardo Lopes Dias, worked for 10 years for New Tribes beside the Javari Valley . New Tribes said its helicopter was now based in Cruzeiro do Sul, 55km south of the reserve.\n“History shows us that any contagious disease can be catastrophic for these people,” said Douglas Rodrigues, a professor of medicine at the Federal University of São Paulo with extensive experience of isolated indigenous groups.\nDoctors also highlighted the danger of Covid-19 for isolated tribes. According to the site Amazônia Real , one adult Marubo indigenous man and his two daughters are suspected of having the virus. All three are reported to be in isolation in Atalaia do Norte, a town on the Amazon’s northern edge. Samples for testing need to be sent to Manaus, over 1,000km away.\n“This virus is a potential killer that could easily wipe out the whole community,” said Adam Mol, a Polish doctor who worked with remote tribes in the Javari Valley.\nMol cited the case of American missionary John Chau , killed by a tribe of hunter-gatherers last year after he landed on North Sentinel Island in the Indian Ocean to illustrate the risks New Tribes present to the region.\n“He literally accepted killing some of them with the excuse of bringing them God’s word. That’s literally how these people think,” Mol said.\nEvangelical Christians have extended their influence in Brazil under the country’s far-right president, Jair Bolsonaro, who increasingly relies on their support and has a history of racist remarks about indigenous people, calling those living on protected reserves “ prehistoric ”.\n“The route is being prepared,” Rodrigues said. “They are ready to invade these territories and do these contacts. New Tribes’ mission is clear.”\nThe indigenous association of the Javari Valley, Univaja, demanded the authorities take action to “avoid the worst”.\nNew Tribes has had a controversial history for decades in Brazil . “They are fundamentalist believers,” said Daniel Everett, an American linguist and former missionary in the Brazilian Amazon who knows the group well. “Their views are extremely 19th century.”\nIn 1991, New Tribes missionaries were expelled by Funai from a remote region where they had contacted an isolated indigenous group of Zo’é indigenous people. In 2014, Warren Scott Kennell was given a 58-year sentence for producing child pornography while working as a missionary for New Tribes in Brazil and admitted sexually abusing children. In 2018, Manoel de Oliveira was handed a three-year prison sentence and fined for keeping Zo’é indigenous people in slave-like conditions on his cashew nut plantation. Prosecutors said he had worked with New Tribes in the region. He is currently appealing.\nBrazil’s Época magazine reported that Ethnos360 had appealed for donations to buy a helicopter to evangelise isolated indigenous groups in western Brazil, near the Peruvian border. This would contravene the policy Funai has had since 1987 of no contact with isolated groups, introduced after tribes were decimated by disease when contact was forced on them.\nNew Tribes pilot Jeremiah Diedrich said in a video that the region where the Javari Valley is located had the highest concentration of uncontacted people groups anywhere in the world. “It is the darkest, densest, hardest to reach place in all South America,” Diedrich said. “This is why we need a helicopter.”\n“The mission leadership is going to be able to assign new teams to reach new people groups in this area,” fellow missionary Julie Diedrich said in the film. In a video shot in front of the Robinson R66 helicopter that was shared on Facebook, New Tribes’ Brasil director, pastor Edward Gomes da Luz, thanked God for its arrival.\nIn February, federal prosecutors sought to suspend Dias’s Funai appointment. In court documents , they quoted a recording of Edward Mantoanelli Luz – a conservative anthropologist, Bolsonaro supporter and son of the New Tribes director, who is not part of the group. In it, he claimed some of the credit for the Dias appointment and vowed the “no contact” policy would end. A judge rejected the prosecutors’ move.\nEdward Mantoanelli Luz told the Guardian that his father was a “great professional” and that Brazil’s policy of no contact had been “manipulated” by other countries. “We have now a new president, a president who looks at Brazil before looking abroad. We need to renew and update this policy,” he said.\nFunai and health professionals could initiate the process of contacting isolated groups, he said. “The evangelical message can come at another moment,” he added, voicing support for Bolsonaro’s policy to allow commercial farming and mining on indigenous reserves, activities currently prohibited.\nLast week, Funai changed its rules to allow contact with isolated groups if Covid-19 made it essential to their survival, and suspended entry to indigenous reserves. Mantoanelli Luz welcomed the move. On Monday, following an outcry, the government effectively changed the rules back again.\nFunai said it has no requests from New Tribes to enter the Javari Valley. Dias declined an interview and did not answer a question on whether the “no contact” policy would be changed.\n“According to the current indigenous policy, the self-determination of isolated peoples is a guaranteed and respected right,” he said in an email.\nThe prosecutors challenging his appointment described an international movement of missionary groups to convert indigenous tribes, including a coalition called Finishing the Task . In a video on the Ethnos360 site called A Vision for Finishing, its CEO in the US, Larry Brown, said : “We can see every tribe reached. It is doable.”\nThe Finishing the Task website lists the Korubo tribe in the Javari Valley as one of its “prioritised unengaged unreached people groups”. Guardian reporters visited their villages in 2018. Three small groups of Korubo were contacted by Funai in recent years in exceptional circumstances, following health and security scares. One remains in isolation.\nDoctor Lucas Albertoni worked on all three contacts. “A common cold could evolve into pneumonia and sepsis in a matter of days without medical assistance,” he said, noting that even contacted Korubo remain vulnerable because immunity takes time to build up. “They are a high-risk group for coronavirus,” he said.\nNew Tribes Mission already has one church in the Javari Valley. Its helicopter will “support mission work in the region”, the group said in a statement. “New Tribes Mission Brazil does not work with isolated people, without contact,” it said.\nBut in 1991, New Tribes missionaries led by Edward Gomes da Luz were expelled by Funai from the eastern Amazon region inhabited by the Zo’é tribe. Funai blamed them for dozens of deaths by disease. Douglas Rodrigues arrived in the region with a Funai team six months after the missionaries had set up their base.\nThe Zo’é tribe were suffering from malaria, which they said they had never seen before the missionaries arrived, and flu. Tests showed none had been vaccinated against anything, Rodrigues said. “It is inadmissible that you enter into contact with a group like this … and don’t vaccinate them,” he said.\n“Without experience in doing contact, the missionaries caused great mortality in these people,” Funai employee Fiorello Parise – who also went there – told anthropologist Felipe Milanez, in his book Memories of Indigenous Specialists (Memórias Sertanistas).\nLuz told BBC Brasil he was the first non-indigenous person to contact the tribe, in 1982. He declined the Guardian’s request for an interview and his group said investigations had proved “the absence of illicit acts” and called the accusations “unfounded and untrue”.\nRodrigues said Covid-19 is a reminder of how dangerous a new virus can be to anyone lacking immunity. In this sense, he said: “We are all wild indigenous people now.”",
            date_updated: 2020-03-23T16:07:36Z,
            date_published: 2020-03-23T07:00:29Z,
        },
    ),
)
